decimal/decimal.nim

```diff
@@ -6,3 +6,93 @@
 # at your option. This file may not be copied, modified, or distributed except according to those terms.
 
 import decimal_lowlevel
+
+type 
+    DecimalType = ptr mpd_t
```

Comment:
We should use a ref ptr mpd_t so that it's tracked by the GC and it's not leaked.
For the high-level API people will forgot to release memory.

Example:

Base type:
https://github.com/mratsim/Arraymancer/blob/686836e8dea45fa3838acaf97c757889e1709fa7/src/tensor/data_structure.nim#L58

Auto-alloc/dealloc
https://github.com/mratsim/Arraymancer/blob/686836e8dea45fa3838acaf97c757889e1709fa7/src/tensor/backend/cuda.nim#L23-L40

Reply:
Have I understood you correctly?: 
We make DecimalType an object and gives it two field, one `Fdata: ptr mpd_t` and one `Fref_tracking: ref[ptr mpd_t]`. `Fdata` is what I called the entire `DecimalType`. `Fref_tracking` is there just to make sure it get GC:ed and it is just an initilized `ref[ptr mpd_t]` and does not point to the same data as `Fdata`.

Reply:
We can even make it a ref[ptr mpd_t] directly.

I used a ref and a raw pointer in Arraymancer to avoid double indirection and because CudaTensor are long-lived.

I don't think it applies in our case:
  - Even if there is a double-indirection the addresses will be hot in cache so the cost is low. In Arraymancer you often do a lot of tensor stuff without actually accessing the data.
  - The mp numbers are short-lived and will probably be created/destroyed often, allocating two fields instead of one in that case is probably not justified.

Reply:
That sounds really good :-D
So all I have to change basically is to cast all return values from mpd procs to DecimalType. And if `d: DecimalType` I access the pointer with `d[]`?

Reply:
And another question: CTX is just an object so it will be GC:ed without having to use `ref`? Or does it matter really, as it lives as long as the program does?

Reply:
I solved it like this:
```nim
type DecimalType = ref[ptr mpd_t]

proc newDecimal*(): DecimalType = 
    new[ptr mpd_t](result)
    result[] = mpd_new(CTX_ADDR)

proc newDecimal*(s: string): DecimalType =
    new[ptr mpd_t](result)
    result[] = mpd_new(CTX_ADDR)
    mpd_set_string(result[], s, CTX_ADDR)

proc `+`*(a, b: DecimalType): DecimalType =
    result = newDecimal()
    mpd_add(result[], a[], b[], CTX_ADDR)
```
It works but I'm not quite sure why it does. Shouldn't `result` already be initilized as a `ref[ptr mpd_t]`? And if so, when I run `new` on it, shouldn't I get a ref of a ref? Lot's of questions from a beginner 😅 I'll have to read up on this a bit more I think. The rest of the high level API shouldn't need much more new knowledge though.

Reply:
This is more idiomatic

```Nim
type DecimalType = ref[ptr mpd_t]

proc newDecimal*(): DecimalType = 
    new result
    result[] = mpd_new(CTX_ADDR)

proc newDecimal*(s: string): DecimalType =
    new result
    result[] = mpd_new(CTX_ADDR)
    mpd_set_string(result[], s, CTX_ADDR)

proc `+`*(a, b: DecimalType): DecimalType =
    result = newDecimal()
    mpd_add(result[], a[], b[], CTX_ADDR)
```

Nim compiler will automatically infer the generic type `ptr mpd_t`.

Here is how it works.
By default Nim initializes types to binary zero, sequences to empty sequences and strings to empty strings.

For ref types, binary zero means a nil pointer.
Also remember that Nim provides an implicit `result` variable.

So just after this:

```Nim
proc newDecimal*(): DecimalType = 
    discard
```

1. You have a `result` of type DecimalType and of value `nil`.
2. You need to initialize it with `new` because the underlying value `ptr mpd_t` needs to be stored somewhere.
3. `new` will do the plumbing: allocate memory, add the GC tracking hooks and initialize the memory location to binary zero as well.
4. Then you access that memory location and assign its value `mpd_new(CTX_ADDR)

In summary, unlike in other language this `new` doesn't create a new instance of an object type. You have your nil object of your desired type, pass it to `new` and it will instantiate it.

There is a [`new` overload that accepts a typedesc](https://nim-lang.org/docs/system.html#new%2Ctypedesc) and can be used as it's common in other language s:

```Nim
proc newDecimal*(): DecimalType = 
    result = new DecimalType
```

but we already have an implicit DecimalType that we now overwrite with new DecimalType, so we do more work.

Reply:
Thank you that cleared up a lot of brain knots 😉👍😅 

So `new(result)` would "make more sense" as `init(result)` for the uninitiated.